docker/tarfile/src.go

```diff
@@ -35,14 +38,63 @@ type layerInfo struct {
 	size int64
 }
 
-// NewSource returns a tarfile.Source for the specified path.
-func NewSource(path string) *Source {
+// NewSource returns a tarfile.Source for the specified path or inputStream.
+// Cannot set path and inputStream at the same time.
+// inputStream is mainly used for the docker-daemon transport.
+func NewSource(path string, inputStream io.ReadCloser) (*Source, error) {
 	// TODO: We could add support for multiple images in a single archive, so
 	//       that people could use docker-archive:opensuse.tar:opensuse:leap as
 	//       the source of an image.
+
+	if path != "" && inputStream != nil {
+		return nil, errors.Errorf("Cannot set both path and inputStream")
```

Comment:
If the two parameters can never be set at the same time, there’s no point in allowing them to be passed simultaneously. It would be cleaner to have a `NewFileSource`/`NewStreamSource` (`NewSourceFromFile`/`NewSourceFromStream`? I don’t care) APIs, and factor the common code, if any, into functions private to this module.

Reply:
Separated to two functions, `NewSourceFromFile` and `NewSourceFromStream`.